cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+     * @param optimizationOptions Options to take into account during optimization.
+     */
+    @Override
+    protected void rebalanceForBroker(Broker broker,
+                                      ClusterModel clusterModel,
+                                      Set<Goal> optimizedGoals,
+                                      OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+        LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+        Map<String, Set<Broker>> brokersByBrokerSet = _brokerSetDataHelper.getBrokersByBrokerSet();
+        Map<Broker, String> brokerSetByBroker = _brokerSetDataHelper.getBrokerSetByBroker();
+
+        String currentBrokerSetId = brokerSetByBroker.get(broker);
+        if (currentBrokerSetId == null) {
+            throw new OptimizationFailureException(String.format("[%s] Cannot find broker set for broker id %s", name(), broker.id()));
+        }
+
+        for (Replica replica : broker.trackedSortedReplicas(replicaSortName(this, false, false))
+                .sortedReplicas(true)) {
+            String originalBrokerBrokerSetId = brokerSetByBroker.get(replica.originalBroker());
+            if (originalBrokerBrokerSetId == null) {
+                throw new OptimizationFailureException(String.format("[%s] Cannot find broker set for broker id %s", name(), replica.originalBroker()
+                        .id()));
+            }
+
+            // If the re-balancing action is performed within the broker set boundary, we are good
+            if (broker.isAlive() && originalBrokerBrokerSetId.equals(currentBrokerSetId)) {
+                continue;
+            }
+            // If the brokerSet awareness condition is violated. Move replica to an eligible broker
+            Set<Broker> eligibleBrokers = brokersByBrokerSet.get(originalBrokerBrokerSetId)
+                    .stream()
+                    .filter(b -> clusterModel.aliveBrokers()
+                            .contains(b))
+                    .collect(Collectors.toSet());
+            if (maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers,
+                                          ActionType.INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals, optimizationOptions) == null) {
+                // If balancing action can not be applied, provide recommendation to add new Brokers.
+                ProvisionRecommendation recommendation = new ProvisionRecommendation.Builder(ProvisionStatus.UNDER_PROVISIONED)
+                        .numBrokers(clusterModel.maxReplicationFactor())
+                        .build();
+                throw new OptimizationFailureException(String.format("[%s] Cannot move replica %s to %s on brokerSet %s",
+                                                                     name(), replica, eligibleBrokers,
+                                                                     originalBrokerBrokerSetId), recommendation);
```

Comment:
Would you be able to explain the intended behavior?

Reply:
For any Broker and it's replicas that rebalance should attempt to move replicas to the original broker set that the replica belonged to. The ``optimize()`` method would call for rebalance and then this logic would fix the current cluster model by moving any non-abiding replicas to the eligible brokers. Eligible brokers are all the alive brokers of the orginal brokerSet of the replica.